parser/parser/codegen/generate_diagnostics.cc

```
@@ -78,6 +78,7 @@ tuple<string, string> MESSAGES[] = {
     {"NoAnonymousBlockArg", "no anonymous block parameter"},
     {"NoAnonymousRestArg", "no anonymous rest parameter"},
     {"NoAnonymousKwrestArg", "no anonymous keyword rest parameter"},
+    {"InvalidKey", "key must be valid as local variables"},
```

I don't know how important this is but both the whitequark parser and the ruby parser seem to report an unexpected token error:

```
❯ ruby-parse -e '{"x":}'
warning: parser/current is loading parser/ruby31, which recognizes 3.1.4-compliant syntax, but you are running 3.1.2.
Please see https://github.com/whitequark/parser#compatibility-with-ruby-mri.
(fragment:0):1:6: error: unexpected token tRCURLY
(fragment:0):1: {"x":}
(fragment:0):1:      ^

~/sorbet (jez-package-namespaces-with-owner)
❯ ruby -e '{"x":}'
-e:1: syntax error, unexpected '}'
{"x":}
```

all else being equal it's probably better to have our own custom error for this if it means that we can provide better parse error recovery for it, so I'm fine with the new dclass but I'm curious to know what's different about our parser vs those parsers that is going through a different code path.

---

Sorry it seems like this is actually specific to pattern matching, and the error message is the same as upstream.

```
~/sorbet (jez-package-namespaces-with-owner)
❯ ruby -e 'case {}; in {"":} then true; end'
-e:1: key must be valid as local variables
case {}; in {"":} then true; end
-e:1: identifier  is not valid to set

~/sorbet (jez-package-namespaces-with-owner)
❯ ruby -e 'case {}; in {"x":} then true; end'
-e:1:in `<main>': {}: key not found: :x (NoMatchingPatternKeyError)

~/sorbet (jez-package-namespaces-with-owner)
❯ ruby -e 'case {}; in {"x y":} then true; end'
-e:1: key must be valid as local variables
case {}; in {"x y":} then true; end
-e:1: identifier x y is not valid to set
```

but in any case, it's not just empty string that's invalid

---

Ah, true. I don't think that we can handle the second case, as it's a dynamic property of the value being examined, but the third case is definitely something we could raise an error for.

---

yep, the second case I was just trying to show that the logic is not "are there quotes around the name" which I suspected might be the case, like maybe `{x:}` was fine but `{"x":}` was not fine because it had quotes (which would be easier to detect in Builder.cc than trying to check valid variable names)

---

Is the plan to merge this and handle the non-empty "key must be valid as local variable" in a follow up change / an open issue?

---

I don't know whether there's an easy helper for this. Might be best to just over-eagerly allow things—I don't think that having `{"x y":}` causes problems for Sorbet.

Might want to add a test case and defer solving in Prism.

---

My timebox ran out, and I agree that this will be easier to handle when we upgrade to prism 👍